### PR TITLE
Add org-cite function

### LINKED ISSUE #108: Add support for new org cite
While it's not done or merged yet, this thread suggests it likely will be "soon" (the syntax is basically done, as is a draft API).

https://lists.gnu.org/archive/html/emacs-orgmode/2021-04/msg00430.html

When it is, citation insertion would need to be adapted to take advantage of it, ideally in bibtex-completion. I don't believe  the `bibtex-completion-insert-citation` function even needs to be changed to do this, given it has customization capabilities; see:

https://github.com/tmalsburg/helm-bibtex#format-of-citations

#113 is adding this.

I'd expect there would be overlap with `org-ref` too.

It seems, however, there will be opportunities here too, regardless of what code ends up where. 

See, for example, #91, which may we the place to start.

PRs welcome, as I'm not sure I have the time or skill to do this ATM.

Aside: this project is under more active current development than bibtex-completion et al. So at times it makes sense to add code here that might later be transferred over to the former later. We use the same license, so that won't be any impediment.

### PR DESCRIPTION
**Pull request by bdarcus**
## Introduction

It looks like `org-cite` is close to merging.

https://lists.gnu.org/archive/html/emacs-orgmode/2021-04/msg00790.html

So this implements a function to format that syntax.

You can use this init file to test this with the org-cite branch code:

https://gist.github.com/bdarcus/2645f99363fc47ddab2aae24c5d9e66c

Here's what the style prompt looks like:

Screenshot from 2021-05-06 09-00-34

... with results when inserting into the `new-org-cite` org branch:

Screenshot from 2021-05-03 13-50-04

## Details

Bibtex-completion has flexible citation formatting functions.

Here is the default currently:

``` elisp
(defcustom bibtex-completion-format-citation-functions
  '((org-mode      . bibtex-completion-format-citation-ebib)
```

So this will hopefully replace `bibtex-completion-format-citation-ebib` as default.

I adapted this from the pandoc function, compared to which I have:

1. changed "prenote" to "prefix" to match `org-cite` language
2. removed "postnote" (because this is mostly used for things like page numbers, which are specific to an individual cite item; users will have to add this manually after insertion)
3. added `org-cite`-specific doc string and examples
4. defined a default style, which is ignored, and others, which are configurable, and which result in `[cite/text:@doe19;@doe20;@doe21]`.

## Issues

In general, I'm a little uncertain on the UX, but I have decided to keep this simple, and have users manually edit the inserted citation for more complex cases.

This is in part related to #91. It might be nice to support this with `completion-at-point`.

### Affixes

`Org-cite` has two-levels:

- one for the citation as a whole
- another for each individual citation within that.

I elected only to initially support prompting for the first, and only for the prefix.

### Styles

The styles currently overlap with those in the `oc-basic` POC, which may end up the fallback.

Regardless, I think merging this may have to wait on andras-simonyi/citeproc-org#16.

Closes #108.

### CODE REVIEW

File: bibtex-actions.el [hunk 1]
```diff
@@ -103,6 +103,31 @@ may be indicated with the same icon but a different face."
...
+  [cite:@key]
+
+... or its 'suppress author' variant:
+
+  [cite:-@key]"
+  (let* ((prenote  (if bibtex-completion-cite-prompt-for-optional-arguments (read-from-minibuffer "Prenote: ") ""))
+         (postnote (if bibtex-completion-cite-prompt-for-optional-arguments (read-from-minibuffer "Postnote: ") ""))
+         (prenote  (if (string= "" prenote)  "" (concat prenote  " ")))
+         (postnote (if (string= "" postnote) "" (concat ", " postnote))))
+    ;; this is derived from the pandoc syntax function, but this has two-levels of affixes
```
Comment (bdarcus) [inline bibtex-actions.el:128]: UX question: does it make sense to prompt users for affixes on each cite, or is that just going to be annoying?

**Comment by bdarcus** [comment]
Here's the completing-read UI for the commands now.

Screenshot from 2021-04-22 18-42-17

File: bibtex-actions.el [hunk 2]
```diff
@@ -103,6 +103,47 @@ may be indicated with the same icon but a different face."
...
+    (format "[cite%s:%s%s%s]" command prenote (s-join ";" (--map (concat "@" it) keys)) postnote)))
+
+(defun bibtex-actions--citation-commands-completion-table (commands)
+  "Create the completion table for COMMANDS."
+  (lambda (string pred action)
+    (if (eq action 'metadata)
+        `(metadata
+          ;; better here to use x-group?
+          (display-sort-function . ,#'identity)
+          (annotation-function . (mapcar (lambda (command) (cdr command) commands)))))
```
Comment (bdarcus) [inline bibtex-actions.el:144]: Not currently working.

File: bibtex-actions.el [hunk 3]
```diff
@@ -103,6 +109,53 @@ may be indicated with the same icon but a different face."
...
+    (format "[cite%s:%s%s]" command prefix (s-join ";" (--map (concat "@" it) keys)))))
+
+(defun bibtex-actions--citation-commands-read (commands default)
+  "Prompt user for citation command/style from COMMANDS, and define DEFAULT."
+  (completing-read
+   "Style: "
+   (lambda (string pred action)
+     (if (eq action 'metadata)
+         '(metadata
+           (affixation-function . (mapcar
```
Comment (bdarcus) [inline bibtex-actions.el:153]: Not currently working.

**Comment by denismaier** [comment]
Let me comment on the current user interface:

First: I think it's totally reasonable to let users edit the
source. In that case you might even want to remove the prefix prompt as
this might confuse some users.

But, if you want to add more support here, what about this:
- user calls bibtex-actions-insert-citation
- minibuffer opens
- ask for a style
- user selects one (!) item, presses enter
- prompt for prefix and suffix for this item
- minibuffer shows the list of items again
- repeat
- Finish the whole process with C-RET

If you think the last part sucks, that could be a distinct
multi-citation function.
So for single item citations, it's:

- user calls bibtex-actions-insert--single-citation
- minibuffer opens
- ask for a style
- user selects item, presses enter
- prompt for prefix and suffix for this item

Global affixes are by far less common then prefixes and suffixes on
individual items in a citation, so I think such an approach could make
sense. Of course, you could still add an option to add global affixes,
perhaps with a prefix function, or a users selects the style with C-RET
or so. That could give you:

- user calls bibtex-actions-insert-citation-with-global-affixes
- minibuffer opens
- ask for a style
- prompt for global prefix
- user selects one (!) item, presses enter
- prompt for prefix and suffix for this item
- minibuffer shows the list of items again
- repeat
- Finish the whole process with C-RET
- prompt for global suffix

**Comment by bdarcus** [comment]
Thanks for the detailed feedback @denismaier!

I think what you suggest is worth experimenting with; hard to know for sure in the abstract.

On this, just to clarify:

> Global affixes are by far less common then prefixes and suffixes on
> individual items in a citation

It's probably confusing, but I was thinking about this for a specific case: "see" references.

I think in that case, they are global; no?

E.g. (for more on foo, see Doe, 2019; Roberts, 2020).

**Comment by denismaier** [comment]
> I think in that case, they are global; no?
> 
> E.g. `(for more on foo, see Doe, 2019; Roberts, 2020)`.
> 
> Edit: and these are pretty common.

Correct. The problem with them is just that you don't really support for that in most backends. Biblatex has support for this. Citeproc-org/citeproc-el, already?

But if you use natbib, relying on a global prefix might give you headaches. OTOH, resolving these issues might be the job of the export framework.

**Comment by bdarcus** [comment]
> Edit: for export to natbib, can just prepend the global prefix to the first cite prefix?

Actually, while you raise this, we should figure out if we want to provide some feedback to Nicolas on this.

Right now, oc-natbib ignores the global prefix. Should it?

Maybe you can think about that and incorporate it into whatever you might post on the org list?

**Comment by denismaier** [comment]
> Edit: for export to natbib, can just prepend the global prefix to the first cite prefix?

That should be possible, yes.

> Right now, oc-natbib ignores the global prefix. Should it?

No, merging would be better.

**Comment by denismaier** [comment]
FWIW, flexible input is as easy as this:

```lisp
(defun bibtex-actions--insert-cite-fragment ()
 "Insert a cite fragment"
  (interactive)
  (let* ((styles bibtex-actions-org-cite-commands)
         (style  (if bibtex-completion-cite-prompt-for-optional-arguments
                     (ido-completing-read "Style: " styles nil nil nil nil "default")))
         (style  (if (string-equal style "default") "" (concat "/" style))))
    (insert (format "[cite%s: " style))))

(defun bibtex-completion-format-citation-default (keys)
  "Default formatter for keys, separates multiple keys in KEYS with commas."
  (s-join ", " (--map (concat "@" it) keys)))
```

- call bibtex-actions--insert-cite-fragment
- choose style
- start typing; global prefix
- add the first citation key with bibtex-actions-insert-cite-key
- add prefix and suffix manually
- repeat
- add global suffix
- close the brackets

**Comment by denismaier** [comment]
> That would be more efficient; no?

Sure. That was just a quick and dirty poor man's solution.

**Comment by bdarcus** [comment]
Thanks for proposing it; it does demonstrate the design challenges I was thinking about.

Ideally, I really want an approach that works really well for both single-cite citations and multi-cite; that provides as little friction as possible for writers.

**Comment by bdarcus** [comment]
> #91 can be used to do a lot of this.

I started #132 to address this.

**Comment by bdarcus** [comment]
> ... and that's it. Anything else is added inline?

I'm a little unclear of dividing line between org completion, and third-party.

In any case, seems for inline there really two capf actions:

1. the citation; e.g. type `[c` and cite completes, with a pop-up for the style.
2. insert key

**Comment by denismaier** [comment]
Just a thought: what's the performance hit of these autocompletion frameworks?

**Comment by bdarcus** [comment]
> Just a thought: what's the performance hit of these autocompletion frameworks?

Likely insignificant.

The capf function I'm playing with would use the same cached completion candidates as the other commands, so should be very performant.

**Comment by bdarcus** [comment]
I updated this to use the "noauthor" instead of "year" style now.

**Comment by denismaier** [comment]
Shouldn't there be a space after the colon?

File: bibtex-actions.el [hunk 4]
```diff
@@ -122,6 +137,40 @@ manager like Zotero or JabRef."
...
+
+  [cite: see ;@key pp23-24]"
+  (let* ((prefix  (if bibtex-completion-cite-prompt-for-optional-arguments (read-from-minibuffer "Prefix: ") ""))
+         (styles bibtex-actions-org-cite-styles)
+         (style  (if bibtex-completion-cite-prompt-for-optional-arguments
+                     (ido-completing-read "Style: " styles nil nil nil nil "default")))
+         (prefix  (if (string= "" prefix)  "" (concat prefix  " ;")))
+         (style  (if (string-equal style "default") "" (concat "/" style))))
+    ;; this is derived from the pandoc syntax function, but this has two-levels of affixes
+    (format "[cite%s:%s%s]" style prefix (s-join ";" (--map (concat "@" it) keys)))))
```
Comment (denismaier) [inline bibtex-actions.el:172]: suggested change:
```elisp
(format "[cite%s: %s%s]" style prefix (s-join ";" (--map (concat "@" it) keys)))))
```

WDYT?

**Comment by bdarcus** [comment]
> Shouldn't there be a space after the colon?

Can you attach the comment to the specific line in question? I'm not sure what you're referring to ;-)

**Comment by denismaier** [comment]
> Can you attach the comment to the specific line in question? I'm not sure what you're referring to ;-)

I've just added an inline comment.

**Comment by denismaier** [comment]
Hmm, there's no space in Nicholas's original post describing the syntax. But I think it's more readable with a space. (And it doesn't seem to affect to export.

File: bibtex-actions.el (hunk 4 above)
Comment (bdarcus) [reply bibtex-actions.el:172]: I updated your comment to use the github "suggested" feature; note the difference in the source. You can access it by clicking the left most icon on the toolbar.

But on the substance, I don't believe so. Wouldn't that result in `[cite: @doe]`?

**Comment by bdarcus** [comment]
> Hmm, there's no space in Nicholas's original post describing the syntax. But I think it's more readable with a space. (And it doesn't seem to affect to export.

Right; the delimiters are online colons and semi-colons. The current behavior is correct, and I think what you're proposing could actually screw things up if there is a prefix.

**Comment by denismaier** [comment]
> Right; the delimiters are only colons and semi-colons. The current behavior is correct, and I think what you're proposing could actually screw things up if there is a prefix.

Could be, yes. Maybe we should check with Nicolas once things are a bit more stable. (Not a high priority issue of course...)

**Comment by bdarcus** [comment]
> Just a thought: what's the performance hit of these autocompletion frameworks?

FYI, @denismaier, I've made enough progress on #91 to confirm performance will not be an issue.

**Pull request merged**